client/client.go

```diff
@@ -66,32 +67,56 @@ func NewClient(rootURL *url.URL, timeout time.Duration) Client {
 }
 
 func NewInsecureClient(rootURL *url.URL, timeout time.Duration) Client {
+	if config.GetJWT() != "" {
+		logger.Fatal("Using JWT on an insecure client is strictly forbidden!")
+	}
 	return newClient(rootURL, timeout, true)
 }
 
 func newClient(rootURL *url.URL, timeout time.Duration, isInsecure bool) Client {
-	baseUrl := getBaseURL(*rootURL)
-
 	client := &client{
-		baseURL: baseUrl,
-		httpClient: &http.Client{
-			Timeout: timeout,
-			Transport: &http.Transport{
-				TLSClientConfig: &tls.Config{InsecureSkipVerify: isInsecure},
-			},
-		},
+		baseURL: getBaseURL(*rootURL),
+		httpClient: newHttpClient(timeout, isInsecure),
 	}
 	client.getApiLinks()
 	return client
 }
 
-
 // getBaseURL creates a url from the url path passed in and the apiVersion
 func getBaseURL(u url.URL) *url.URL {
 	u.Path = path.Join(u.Path, ApiVersion)
 	return &u
 }
 
+func newHttpClient(timeout time.Duration, isInsecure bool) *http.Client {
+	httpClient := &http.Client{
+		Timeout: timeout,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: isInsecure},
+		},
+	}
+
+	// this decorates the client transport with the jwt header
+	if config.GetJWT() != "" {
+		t := withHeader{Header: make(http.Header), rt: httpClient.Transport}
+		t.Set("Authorization", fmt.Sprintf("Bearer %s", config.GetJWT()))
+		httpClient.Transport = t
+	}
+	return httpClient
+}
+
+type withHeader struct {
```

Comment:
withHeader but what it is? Seems like there is a missing word e.g. Transport or RoundTripper

In my opinion I think this goes well beyond of actually what is required.

```type jwtTransport struct {
	jwt string
	rt http.RoundTripper
}

func newJwtTransport(rt http.RoundTripper, jwt string) *jwtTransport {
	if rt == nil {
		panic("RoundTripper must be provided")
	}
	return &jwtTransport{
		jwt: jwt,
		rt: rt,
	}
}

func (t jwtTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", config.GetJWT()))
	return t.rt.RoundTrip(req)
}```

Reply:
ok renamed the method after our conversation